zerver/lib/topic.py

```diff
@@ -321,3 +321,38 @@ def participants_for_topic(realm_id: int, recipient_id: int, topic_name: str) ->
         ).values_list("id", flat=True)
     )
     return participants
+
+def get_topic_creator_user_id(realm_id: int, recipient_id: int, topic_name: str) -> int | None:
+    """
+    User who created the topic. It is usefull for support streams.
+    """
+    first_message_in_topic = Message.objects.filter(
+        # Uses index: zerver_message_realm_recipient_upper_subject
+        realm_id=realm_id,
+        recipient_id=recipient_id,
+        subject__iexact=topic_name,
+    ).order_by('id').first()
+
+    if first_message_in_topic is None:
+        return None
+    
+    return first_message_in_topic.sender_id
+
+def check_access_based_on_stream_topic_view_policy(sender: UserProfile, stream: Stream) -> bool:
```

Comment:
This should become a group membership check with the new groups-based permissions systems, which should be much cheaper and also a lot less semi-duplicated code to write.